### DISCUSSION TITLE
Passing additional parameters to create object

### DISCUSSION THREAD

**Discussion by phaza** [riok/mapperly#447] [Q&A]
I love the idea of mapperly, it hands down has the best approach to mapping and mostly does what I need.
However, I've shit a snag:

```csharp
public record RecordA()
{
    public required long PropA { get; init; }
    public required long PropB { get; init; }
}
public record RecordB(long ParamC)
{
    public required long PropA { get; init; }
    public required long PropB { get; init; }
}

[Mapper]
public partial class Map
{    
    public partial RecordB ToRecordB((RecordA A, long ParamC) input);
    public partial RecordB ToRecordB(RecordA A, long ParamC); // Unsupported method signature
}
```

Is there any way I can get this working?
Mapperly doesn't seem to like named tuples as it complains about "Item1" and "Item2" amongst other things.
I know if the properties weren't init props I could write a void mapping method, but it feels wrong to change my classes to support mapperly.

If anyone wonders about the use case it's an aspnetcore application where part of the parameters come from the url, while the rest are provided in the body, so passing them all as one model to the domain requires this mapping.

**Comment by phaza** [comment]
Mapping the properties like this almost works, but 1. is very tedious and 2. doesn't work because I get RMG020: "The member Item1 on mapping source etc.)"

```csharp
[Mapper]
public static partial class Map
{
    [MapProperty(new[] { nameof(input.A), nameof(input.A.PropA) }, new[] { nameof(RecordB.PropA) })]
    [MapProperty(new[] { nameof(input.A), nameof(input.A.PropB) }, new[] { nameof(RecordB.PropB) })]
    public static partial RecordB ToRecordB(this (RecordA A, long ParamC) input);
}
```

result:
```csharp
    public static partial class Map
    {
        public static partial global::RecordB ToRecordB(this (global::RecordA A, long ParamC) input)
        {
            var target = new global::RecordB(input.ParamC)
            {
                PropA = input.A.PropA,
                PropB = input.A.PropB
            };
            return target;
        }
    }
```

**Comment by TimothyMakkison** [comment]
Unfortunately mapperly does not support multiple mapping parameters 😞, although this is likely to change in the future. See #103. 

Your manual work around is really clever, I'm not sure why mapperly can access both your named field aliases and the default tuple field names, I wonder if this is due to a roslyn quirk. Mapperly records which members have been mapped and which haven't. In your case mapperly sees your source as having 4 members: `RecordA A`, `long ParamC`, `RecordA Item1` and `long Item2`. `A` and `ParamC` are mapped but Mapperly thinks that `Item1` and `Item2` have not been mapped so emits an error.

I'll see if I can prevent RMG020 when the field has been mapped.

**Reply by phaza** [reply]
Thanks, really love your work!

Another thing which would be awesome here is if I could say "map all properies from input.A to RecordB, instead of having to map them all one by one. Would simplify flattening a lot.

**Reply by TimothyMakkison** [reply]
Yeah I'm not sure if Mapperly lets you do `[MapProperty(nameof(input.A), "")]`. I don't think this would be needed if multiple input parameters were supported?

Mapperly does infer member paths for flattening, so the following works:
```C#
public record RecordA()
{
    public required long PropA { get; init; }
    public required long PropB { get; init; }
}
public record RecordB(long ParamC)
{
    public required long APropA { get; init; }
    public required long APropB { get; init; }
}

[Mapper]
public static partial class Map
{
    public static partial RecordB ToRecordB(this (RecordA A, long ParamC) input);
}
```

**Reply by phaza** [reply]
Yeah, I know I can getting flattening working that way, but then I have to rename properties to something that doesn't really make sense to me. And sometimes you don't necessarily have control of the class you're converting to.  

Another use case to allow `[MapProperty(nameof(input.A), "")]` is merging of multiple objects into a single object.

**Reply by latonz** [reply]
Created https://github.com/riok/mapperly/issues/453 to add a configuration attribute to include several flattened members at once if the name match.  This can then be used to merge objects.

**Reply by latonz** [reply]
Closing as this is discussed in #453
